types/wlr_input_method_v2.c

```diff
@@ -108,10 +116,126 @@ static void im_delete_surrounding_text(struct wl_client *client,
 	input_method->pending.delete.after_length = after_length;
 }
 
+bool wlr_surface_is_input_popup_surface_v2(struct wlr_surface *surface) {
+	return surface->role == &input_popup_surface_v2_role;
+}
+
+struct wlr_input_popup_surface_v2 *wlr_input_popup_surface_v2_from_wlr_surface(
+		struct wlr_surface *surface) {
+	assert(wlr_surface_is_input_popup_surface_v2(surface));
+	return (struct wlr_input_popup_surface_v2 *)surface->role_data;
+}
+
+void wlr_input_popup_surface_v2_send_text_input_rectangle(
+		struct wlr_input_popup_surface_v2 *popup_surface,
+		struct wlr_box *sbox) {
+	zwp_input_popup_surface_v2_send_text_input_rectangle(
+		popup_surface->resource, sbox->x, sbox->y, sbox->width, sbox->height);
+}
+
+static void popup_surface_set_mapped(
+		struct wlr_input_popup_surface_v2 *popup_surface, bool mapped) {
+	if (mapped && !popup_surface->mapped) {
+		popup_surface->mapped = true;
+		wlr_signal_emit_safe(&popup_surface->events.map, popup_surface);
+	} else if (!mapped && popup_surface->mapped) {
+		wlr_signal_emit_safe(&popup_surface->events.unmap, popup_surface);
+		popup_surface->mapped = false;
+	}
+}
+
+static void popup_surface_surface_role_precommit(struct wlr_surface *surface) {
+	struct wlr_input_popup_surface_v2 *popup_surface = surface->role_data;
+	if (popup_surface == NULL) {
+		return;
+	}
+	if (surface->pending.committed & WLR_SURFACE_STATE_BUFFER &&
+			surface->pending.buffer_resource == NULL) {
```

Comment:
Is there a reason we need to handle this case separately? Isn't this already handled in `popup_surface_surface_role_commit`?